In [1]:
#importing the libraries

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import sklearn
from sklearn import model_selection
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix


pd.options.display.max_columns=1000
pd.options.display.max_rows=1000
from IPython.display import display
#importing the dataset
dataset = pd.read_csv('NSL_KDD-master/KDDTrain+.txt',header=None)
dataset = dataset.iloc[:, :-1]
display(dataset.sample(n=20))

0     1         2    3     4      5   6   7   8   9   10  11  12  13  \
29060    0   tcp      http   SF   192    320   0   0   0   0   0   1   0   0   
25725    0   tcp      smtp   S0     0      0   0   0   0   0   0   0   0   0   
78399    0   tcp      http   SF   343    255   0   0   0   0   0   1   0   0   
7188     0   tcp      http   SF   147  38988   0   0   0   1   0   1   0   0   
75000    0   tcp      http   SF   251  28066   0   0   0   0   0   1   0   0   
112816   0   tcp   private   S0     0      0   0   0   0   0   0   0   0   0   
10717    0   tcp      http   SF   301   1517   0   0   0   0   0   1   0   0   
61498    0   tcp  http_443   S0     0      0   0   0   0   0   0   0   0   0   
1346     0   udp  domain_u   SF    43     71   0   0   0   0   0   0   0   0   
106687   0  icmp     eco_i   SF    18      0   0   0   0   0   0   0   0   0   
62740    0   tcp   private   S0     0      0   0   0   0   0   0   0   0   0   
95283    0   tcp      http   SF   240    311   0   0   0   0   0   1   0   0   
35397    0  icmp     eco_i   SF     8      0   0   0   0   0   0   0   0   0   
18229    0   tcp      smtp   SF  1049    326   0   0   0   0   0   1   0   0   
78225    0   tcp     login   S0     0      0   0   0   0   0   0   0   0   0   
124054   0   tcp  iso_tsap   S0     0      0   0   0   0   0   0   0   0   0   
95888    0   udp  domain_u   SF    34      0   0   0   0   0   0   0   0   0   
80723    0   tcp    domain  REJ     0      0   0   0   0   0   0   0   0   0   
89752    0   tcp   private   S0     0      0   0   0   0   0   0   0   0   0   
40522    0   tcp      http   SF   158   2024   0   0   0   0   0   1   0   0   

        14  15  16  17  18  19  20  21   22   23   24   25   26   27    28  \
29060    0   0   0   0   0   0   0   0   23   24  0.0  0.0  0.0  0.0  1.00   
25725    0   0   0   0   0   0   0   0  104   24  1.0  1.0  0.0  0.0  0.23   
78399    0   0   0   0   0   0   0   0   27   32  0.0  0.0  0.0  0.0  1.00   
7188     0   0   0   0   0   0   0   0    1    2  0.0  0.0  0.0  0.0  1.00   
75000    0   0   0   0   0   0   0   0    1    1  0.0  0.0  0.0  0.0  1.00   
112816   0   0   0   0   0   0   0   0  244   12  1.0  1.0  0.0  0.0  0.05   
10717    0   0   0   0   0   0   0   0    3    4  0.0  0.0  0.0  0.0  1.00   
61498    0   0   0   0   0   0   0   0    2    1  1.0  1.0  0.0  0.0  0.50   
1346     0   0   0   0   0   0   0   0  266  265  0.0  0.0  0.0  0.0  1.00   
106687   0   0   0   0   0   0   0   0    1    1  0.0  0.0  0.0  0.0  1.00   
62740    0   0   0   0   0   0   0   0  227   25  1.0  1.0  0.0  0.0  0.11   
95283    0   0   0   0   0   0   0   0    4   10  0.0  0.0  0.0  0.0  1.00   
35397    0   0   0   0   0   0   0   0    1   44  0.0  0.0  0.0  0.0  1.00   
18229    0   0   0   0   0   0   0   0    1    1  0.0  0.0  0.0  0.0  1.00   
78225    0   0   0   0   0   0   0   0  140    7  1.0  1.0  0.0  0.0  0.05   
124054   0   0   0   0   0   0   0   0  193   12  1.0  1.0  0.0  0.0  0.06   
95888    0   0   0   0   0   0   0   0    1    2  0.0  0.0  0.0  0.0  1.00   
80723    0   0   0   0   0   0   0   0  265   19  0.0  0.0  1.0  1.0  0.07   
89752    0   0   0   0   0   0   0   0   56   19  1.0  1.0  0.0  0.0  0.34   
40522    0   0   0   0   0   0   0   0    2    2  0.0  0.0  0.0  0.0  1.00   

          29    30   31   32    33    34    35    36   37    38   39   40  \
29060   0.00  0.08   64  255  1.00  0.00  0.02  0.02  0.0  0.00  0.0  0.0   
25725   0.06  0.00  255    6  0.02  0.06  0.00  0.00  1.0  1.00  0.0  0.0   
78399   0.00  0.06   27  255  1.00  0.00  0.04  0.06  0.0  0.00  0.0  0.0   
7188    0.00  1.00   47  255  1.00  0.00  0.02  0.02  0.0  0.00  0.0  0.0   
75000   0.00  0.00   25  255  1.00  0.00  0.04  0.05  0.0  0.00  0.0  0.0   
112816  0.06  0.00  255    9  0.04  0.06  0.00  0.00  1.0  1.00  0.0  0.0   
10717   0.00  0.50    9  255  1.00  0.00  0.11  0.04  0.0  0.01  0.0  0.0   
61498   1.00  0.00  255   11  0.04  0.09  0.00  0.00  1.0  1.00  0.0  0.

In [2]:
#Checking the different categories of each features availbale
print(dataset.shape)
#display((set(dataset['tcp'])))
display(len(set(dataset[1])))

#display(set(dataset['http']))
display(len(set(dataset[2])))

display(len(set(dataset[3])))

display(len(set(dataset[41])))

(125973, 42)


3

70

11

23

In [3]:
#replace all other types mentioned as attack.
dataset[41] = dataset[41].replace(['back', 'buffer_overflow', 'ftp_write', 'guess_passwd', 'imap', 'ipsweep', 'land', 'loadmodule', 'multihop', 'neptune', 'nmap', 'perl', 'phf', 'pod', 'portsweep', 'rootkit', 'satan', 'smurf', 'spy', 'teardrop', 'warezclient', 'warezmaster'], 'attack')
display(dataset.sample(10))

x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 41].values

0    1         2     3    4      5   6   7   8   9   10  11  12  13  \
48801    0  tcp   private   REJ    0      0   0   0   0   0   0   0   0   0   
13339    0  tcp      http    S0    0      0   0   0   0   0   0   0   0   0   
108919   0  tcp   printer    S0    0      0   0   0   0   0   0   0   0   0   
117564   0  tcp      smtp    SF  756    335   0   0   0   0   0   1   0   0   
84856    0  tcp      http    SF  324  11972   0   0   0   0   0   1   0   0   
57270    0  tcp   private    S0    0      0   0   0   0   0   0   0   0   0   
98355    0  tcp    Z39_50   REJ    0      0   0   0   0   0   0   0   0   0   
34190    0  tcp    telnet  RSTO    0      0   0   0   0   0   0   0   0   0   
74075    0  tcp  ftp_data    SF  334      0   0   0   0   0   0   1   0   0   
64311    0  tcp    klogin    S0    0      0   0   0   0   0   0   0   0   0   

        14  15  16  17  18  19  20  21   22  23    24    25   26    27    28  \
48801    0   0   0   0   0   0   0   0  171   7  0.00  0.00  1.0  1.00  0.04   
13339    0   0   0   0   0   0   0   0   12  13  1.00  0.92  0.0  0.08  1.00   
108919   0   0   0   0   0   0   0   0  122   7  1.00  1.00  0.0  0.00  0.06   
117564   0   0   0   0   0   0   0   0    1   2  0.00  0.00  0.0  0.00  1.00   
84856    0   0   0   0   0   0   0   0    2   2  0.00  0.00  0.0  0.00  1.00   
57270    0   0   0   0   0   0   0   0  123  16  0.98  1.00  0.0  0.00  0.13   
98355    0   0   0   0   0   0   0   0  208   4  0.00  0.00  1.0  1.00  0.02   
34190    0   0   0   0   0   0   0   0   17   1  0.00  0.00  1.0  1.00  0.06   
74075    0   0   0   0   0   0   0   0    2   2  0.00  0.00  0.0  0.00  1.00   
64311    0   0   0   0   0   0   0   0  127   3  1.00  1.00  0.0  0.00  0.02   

          29    30   31   32    33    34    35    36   37   38    39   40  \
48801   0.08  0.00  255    7  0.03  0.08  0.00  0.00  0.0  0.0  1.00  1.0   
13339   0.00  0.15   12   12  1.00  0.00  0.08  0.00  1.0  1.0  0.00  0.0   
108919  0.06  0.00  255    7  0.03  0.07  0.00  0.00  1.0  1.0  0.00  0.0   
117564  0.00  1.00   63  121  0.54  0.08  0.02  0.02  0.0  0.0  0.00  0.0   
84856   0.00  0.00   51  255  1.00  0.00  0.02  0.06  0.0  0.0  0.00  0.0   
57270   0.07  0.00  255   14  0.05  0.05  0.00  0.00  1.0  1.0  0.00  0.0   
98355   0.07  0.00  255    4  0.02  0.08  0.00  0.00  0.0  0.0  1.00  1.0   
34190   0.59  0.00  255    2  0.01  0.05  0.07  0.00  0.0  0.0  0.85  0.5   
74075   0.00  0.00    1    1  1.00  0.00  1.00  0.00  0.0  0.0  0.00  0.0   
64311   0.07  0.00  255    3  0.01  0.06  0.00  0.00  1.0  1.0  0.00  0.0   

            41  
48801   attack  
13339   attack  
108919  attack  
117564  normal  
84856   normal  
57270   attack  
98355   attack  
34190   attack  
74075   normal  
64311   attack

In [4]:
#encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

#first applying label encoding to convert strings to number.
labelencoder_x_1 = LabelEncoder()
labelencoder_x_2 = LabelEncoder()
labelencoder_x_3 = LabelEncoder()
x[:, 1] = labelencoder_x_1.fit_transform(x[:, 1])
x[:, 2] = labelencoder_x_2.fit_transform(x[:, 2])
x[:, 3] = labelencoder_x_3.fit_transform(x[:, 3])

display(set(x[:,1]))# {0, 1, 2}= 3 categories
#display(set(x[:,2]))# {0,1,2,3.....69}= 70 categories
display(set(x[:,3]))# {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}= 11 categories

{0, 1, 2}

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

In [5]:
# Now applying one-hot-encoding to map the numbers to appropiate weights.
onehotencoder_1 = OneHotEncoder(categorical_features = [1])
x = onehotencoder_1.fit_transform(x).toarray()
onehotencoder_2 = OneHotEncoder(categorical_features = [4])
x = onehotencoder_2.fit_transform(x).toarray()
onehotencoder_3 = OneHotEncoder(categorical_features = [74]) #4+70-1
x = onehotencoder_3.fit_transform(x).toarray()

labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)
#display(y)
display(x.shape)

(125973, 122)

# splitting the dataset into the training set and test set

#from sklearn.cross_validation import train_test_split
#x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)
#spliutting in 70% and 30%



In [6]:
# Now preparing the train set.
x_train= x
y_train = y
display(y_train.shape)

(125973,)

In [7]:
#Preparing the test set.

test_data=pd.read_csv('./NSL_KDD-master/test.csv')
#display(test_data.head(5))
x_test= test_data.iloc[:, :-1]
#display(x_test.head())

x_test=x_test.values
display((x_test.shape))

y_test= test_data.iloc[:,122]
y_test = y_test.values
display((y_test.shape))

(22550, 122)

(22550,)

In [8]:
#feature scaling of the dataset.
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)

df2 = pd.DataFrame(x_test)
display(df2.head())
display(x_train)

0         1         2         3         4         5         6    \
0 -0.019113  3.196020 -0.112050 -0.028606 -0.139982 -0.618438 -0.053906   
1 -0.019113  3.196020 -0.112050 -0.028606 -0.139982 -0.618438 -0.053906   
2 -0.019113 -0.312889 -0.112050 -0.028606 -0.139982 -0.618438 -0.053906   
3 -0.019113 -0.312889 -0.112050 -0.028606 -0.139982 -0.618438 -0.053906   
4 -0.019113 -0.312889  8.924602 -0.028606 -0.139982 -0.618438 -0.053906   

        7         8         9         10        11       12        13   \
0 -0.031767 -0.019726 -1.211901 -0.046432 -0.038557 -0.02408 -0.083005   
1 -0.031767 -0.019726 -1.211901 -0.046432 -0.038557 -0.02408 -0.083005   
2 -0.031767 -0.019726  0.825150 -0.046432 -0.038557 -0.02408 -0.083005   
3 -0.031767 -0.019726  0.825150 -0.046432 -0.038557 -0.02408 -0.083005   
4 -0.031767 -0.019726 -1.211901 -0.046432 -0.038557 -0.02408 -0.083005   

        14        15        16        17        18        19        20   \
0 -0.003985 -0.087401 -0.075287 -0.076556 -0.065918 -0.067002 -0.064444   
1 -0.003985 -0.087401 -0.075287 -0.076556 -0.065918 -0.067002 -0.064444   
2 -0.003985 -0.087401 -0.075287 -0.076556 -0.065918 -0.067002 -0.064444   
3 -0.003985 -0.087401 -0.075287 -0.076556 -0.065918 -0.067002 -0.064444   
4 -0.003985 -0.087401 -0.075287 -0.076556 -0.065918 -0.067002 -0.064444   

        21       22        23        24        25        26        27   \
0 -0.065491 -0.06736 -0.278095 -0.058797 -0.194371 -0.158232 -0.062168   
1 -0.065491 -0.06736 -0.278095 -0.058797 -0.194371 -0.158232 -0.062168   
2 -0.065491 -0.06736 -0.278095 -0.058797 -0.194371 -0.158232 -0.062168   
3 -0.065491 -0.06736 -0.278095 -0.058797  5.144807 -0.158232 -0.062168   
4 -0.065491 -0.06736 -0.278095 -0.058797 -0.194371 -0.158232 -0.062168   

        28        29        30        31        32        33        34   \
0 -0.061457 -0.119274 -0.118829 -0.239984 -0.064257 -0.003985 -0.060539   
1 -0.061457 -0.119274 -0.118829 -0.239984 -0.064257 -0.003985 -0.060539   
2 -0.061457 -0.119274 -0.118829  4.166943 -0.064257 -0.003985 -0.060539   
3 -0.061457 -0.119274 -0.118829 -0.239984 -0.064257 -0.003985 -0.060539   
4 -0.061457 -0.119274 -0.118829 -0.239984 -0.064257 -0.003985 -0.060539   

        35        36     37        38        39       40        41        42   \
0 -0.686328 -0.002817 -0.065 -0.003985 -0.071851 -0.07405 -0.058729 -0.048777   
1 -0.686328 -0.002817 -0.065 -0.003985 -0.071851 -0.07405 -0.058729 -0.048777   
2 -0.686328 -0.002817 -0.065 -0.003985 -0.071851 -0.07405 -0.058729 -0.048777   
3 -0.686328 -0.002817 -0.065 -0.003985 -0.071851 -0.07405 -0.058729 -0.048777   
4 -0.686328 -0.002817 -0.065 -0.003985 -0.071851 -0.07405 -0.058729 -0.048777   

        43        44        45        46        47        48        49   \
0 -0.057143 -0.061522 -0.058456 -0.059136 -0.059942 -0.056792 -0.052556   
1 -0.057143 -0.061522 -0.058456 -0.059136 -0.059942 -0.056792 -0.052556   
2 -0.057143 -0.061522 -0.058456 -0.059136 -0.059942 -0.056792 -0.052556   
3 -0.057143 -0.061522 -0.058456 -0.059136 -0.059942 -0.056792 -0.052556   
4 -0.057143 -0.061522 -0.058456 -0.059136 -0.059942 -0.056792 -0.052556   

        50        51        52        53        54        55      56   \
0 -0.053683 -0.053535 -0.070896 -0.048531 -0.036543 -0.189322 -0.0063   
1 -0.053683 -0.053535 -0.070896 -0.048531 -0.036543 -0.189322 -0.0063   
2 -0.053683 -0.053535 -0.070896 -0.048531 -0.036543 -0.189322 -0.0063   
3 -0.053683 -0.053535 -0.070896 -0.048531 -0.036543 -0.189322 -0.0063   
4 -0.053683 -0.053535 -0.070896 -0.048531 -0.036543 -0.189322 -0.0063   

        57        58       59        60        61        62        63   \
0 -0.024891 -0.045827 -0.02341  2.182788 -0.007969 -0.024891 -0.026137   
1 -0.024891 -0.045827 -0.02341  2.182788 -0.007969 -0.024891 -0.026137   
2 -0.024891 -0.045827 -0.02341 -0.458130 -0.007969 -0.024891 -0.026137   
3 -0.024891 -0.045827 -0.02341 -0.458130 -0.007969 -0.024891 -0.026137   
4 -0.024891 -0.045

array([[-0.01911259, -0.31288912, -0.11204981, ..., -0.6248708 ,
        -0.22453241, -0.37638703],
       [-0.01911259, -0.31288912, -0.11204981, ..., -0.6248708 ,
        -0.38763462, -0.37638703],
       [-0.01911259, -0.31288912, -0.11204981, ...,  1.6189552 ,
        -0.38763462, -0.37638703],
       ...,
       [-0.01911259, -0.31288912, -0.11204981, ..., -0.6248708 ,
        -0.35501418, -0.37638703],
       [-0.01911259, -0.31288912, -0.11204981, ...,  1.6189552 ,
        -0.38763462, -0.37638703],
       [-0.01911259, -0.31288912, -0.11204981, ..., -0.6248708 ,
        -0.38763462, -0.37638703]])

In [99]:
# TAking Decision Tree as classifier, and testing on the model
classifier = DecisionTreeClassifier()
classifier.fit(x_train,y_train)

# Predicting the Test set results
y_pred = classifier.predict(x_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
display(cm)
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = x_train, y = y_train, cv = 10)
accuracies.mean()
accuracies.std()

print()
print("For Decision Tree The Performance result is: ")
print()

#the performance of the classification model
#true negatives is C_{0,0}, 
#false negatives is C_{1,0}, 
#true positives is C_{1,1}
#false positives is C_{0,1}.
print("the Accuracy is: "+ str((cm[0,0]+cm[1,1])/(cm[0,0]+cm[0,1]+cm[1,0]+cm[1,1])))
print("True Positive Rate: "+ str(cm[1,1]/(cm[1,1]+cm[1,0])))
recall=cm[1,1]/(cm[1,0]+cm[1,1])
print("Recall is : "+ str(recall))
print("False Positive rate: "+ str(cm[0,1]/(cm[0,1]+cm[0,0])))
precision= cm[1,1]/(cm[1,0]+cm[0,1])
print("Precision is: "+ str(precision))
print("F-measure is: "+ str(2*((precision*recall)/(precision+recall))))
from math import log
print("Entropy is: "+ str(-precision*log(precision)))


array([[8644, 4196],
       [ 375, 9335]], dtype=int64)


For Decision Tree The Performance result is: 

the Accuracy is: 0.7972949002217294
True Positive Rate: 0.9613800205973223
Recall is : 0.9613800205973223
False Positive rate: 0.326791277258567


TypeError: 'precision' is an invalid keyword argument for this function

In [ ]:
# Checking the number of correct predictions
result=pd.DataFrame(y_pred)
result.insert(loc=1,value=y_test,column='test output')
diff= (result['test output'])-(result[0])
#display(diff)
type(diff)
result.insert(loc=2,value=diff,column='Result')
#display(result)
Validity = []
for i in result['Result']:
    if(i== -1):
        #print('Incorrect Prediction')
        Validity.append('Incorrect Prediction')
    else:
        #print('Correctly Predicted')V
        Validity.append('Correct Prediction')
result.insert(loc=3,value=Validity,column='Validity')
display(result)

In [105]:
#Adding validity column to csv file
final_prediction=pd.read_csv('./NSL_KDD-master/test.csv')
final_prediction.insert(loc=123,value=Validity,column='Validity')
final_prediction.shape
final_prediction.to_csv('Predicted_Data.csv',index=False,index_label=False)

(22550, 124)

In [ ]:
models = []
models.append(('Logistic Regression', LogisticRegression()))
models.append(('Linear Discriminant Analysis', LinearDiscriminantAnalysis()))
#models.append(('K-Nearest Neighbour ', KNeighborsClassifier()))
models.append(('Decison Tree', DecisionTreeClassifier()))
models.append(('Guassian Naive Bayes', GaussianNB()))
#models.append(('SVM', SVC()))

for name,model in models:
    
    # Fitting Naive Bayes to the Training set
    classifier = model
    classifier.fit(x_train, y_train)

    # Predicting the Test set results
    y_pred = classifier.predict(x_test)

    # Making the Confusion Matrix
    from sklearn.metrics import confusion_matrix
    cm = confusion_matrix(y_test, y_pred)

    # Applying k-Fold Cross Validation
    from sklearn.model_selection import cross_val_score
    accuracies = cross_val_score(estimator = classifier, X = x_train, y = y_train, cv = 10)
    accuracies.mean()
    accuracies.std()
    
    print()
    print("For {0} The Performance result is: ".format(name))
    print()

    #the performance of the classification model
    #true negatives is C_{0,0}, 
    #false negatives is C_{1,0}, 
    #true positives is C_{1,1}
    #false positives is C_{0,1}.
    print("the Accuracy is: "+ str((cm[0,0]+cm[1,1])/(cm[0,0]+cm[0,1]+cm[1,0]+cm[1,1])))
    print("True Positive Rate: "+ str(cm[1,1]/(cm[1,1]+cm[1,0])))
    recall=cm[1,1]/(cm[1,0]+cm[1,1])
    print("Recall is : "+ str(recall))
    print("False Positive rate: "+ str(cm[0,1]/(cm[0,1]+cm[0,0])))
    precision= cm[1,1]/(cm[1,0]+cm[0,1])
    print("Precision is: "+ str(precision))
    print("F-measure is: "+ str(2*((precision*recall)/(precision+recall))))
    from math import log
    print("Entropy is: "+ str(-precision*log(precision)))





For Logistic Regression The Performance result is: 

the Accuracy is: 0.7510421286031043
Recall is : 0.6475079227888216
False Positive rate: 0.0830833141010847
Precision is: 0.925849639546859
F-measure is: 0.7620581503772146
Entropy is: 0.07133063543142577


C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")
C:\Users\admin\Anacond


For Linear Discriminant Analysis The Performance result is: 

the Accuracy is: 0.7615964523281596
Recall is : 0.6573939570017432
False Positive rate: 0.07503985424732407
Precision is: 0.9321318228630278
F-measure is: 0.7710196779964221
Entropy is: 0.06551118783008927

For Decison Tree The Performance result is: 

the Accuracy is: 0.8113082039911308
Recall is : 0.7039862388751776
False Positive rate: 0.03235646584595272
Precision is: 0.9694129763130793
F-measure is: 0.8156492353017635
Entropy is: 0.030114397007745534

For Guassian Naive Bayes The Performance result is: 

the Accuracy is: 0.5546341463414635
Recall is : 0.4915211081122371
False Positive rate: 0.01991074493649159
Precision is: 0.9940267765190525
F-measure is: 0.6577844413398303
Entropy is: 0.005955348154933777


For Logistic Regression The Performance result is: 

the Accuracy is: 0.7510421286031043
Recall is : 0.6475079227888216
False Positive rate: 0.0830833141010847
Precision is: 0.925849639546859
F-measure is: 0.7620581503772146
Entropy is: 0.07133063543142577
    

For Linear Discriminant Analysis The Performance result is: 

the Accuracy is: 0.7615964523281596
Recall is : 0.6573939570017432
False Positive rate: 0.07503985424732407
Precision is: 0.9321318228630278
F-measure is: 0.7710196779964221
Entropy is: 0.06551118783008927

For Decison Tree The Performance result is: 

the Accuracy is: 0.8113082039911308
Recall is : 0.7039862388751776
False Positive rate: 0.03235646584595272
Precision is: 0.9694129763130793
F-measure is: 0.8156492353017635
Entropy is: 0.030114397007745534

For Guassian Naive Bayes The Performance result is: 

the Accuracy is: 0.5546341463414635
Recall is : 0.4915211081122371
False Positive rate: 0.01991074493649159
Precision is: 0.9940267765190525
F-measure is: 0.6577844413398303
Entropy is: 0.005955348154933777

In [103]:
#MLP Classifier
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(solver='lbfgs', alpha=1e-5,hidden_layer_sizes=(5, 2), random_state=1)
classifier.fit(x_train,y_train)
y_pred = classifier.predict(x_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# Applying k-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = MLPClassifier(), X = x_train, y = y_train, cv = 10)
accuracies.mean()
accuracies.std()

#the performance of the classification model
#true negatives is C_{0,0}, 
#false negatives is C_{1,0}, 
#true positives is C_{1,1}
#false positives is C_{0,1}.
print("the Accuracy is: "+ str((cm[0,0]+cm[1,1])/(cm[0,0]+cm[0,1]+cm[1,0]+cm[1,1])))
print("True Positive Rate: "+ str(cm[1,1]/(cm[1,1]+cm[1,0])))
recall=cm[1,1]/(cm[1,0]+cm[1,1])
print("Recall is : "+ str(recall))
print("False Positive rate: "+ str(cm[0,1]/(cm[0,1]+cm[0,0])))
precision= cm[1,1]/(cm[1,0]+cm[0,1])
print("Precision is: "+ str(precision))
print("F-measure is: "+ str(2*((precision*recall)/(precision+recall))))
from math import log
print("Entropy is: "+ str(-precision*log(precision)))


the Accuracy is: 0.7549445676274945
True Positive Rate: 0.9218331616889804
Recall is : 0.9218331616889804
False Positive rate: 0.37126168224299066
Precision is: 1.6197973217517192
F-measure is: 1.1749803097925964
Entropy is: -0.7812299189196893
